aten/src/ATen/native/cuda/MaxUnpooling.cu

```diff
@@ -0,0 +1,598 @@
+    const int64_t inputHeight,
+    const int64_t inputWidth,
+    const int64_t outputHeight,
+    const int64_t outputWidth,
+    T* output) {
+  for (int linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < numInputElements;
+       linearIndex += blockDim.x * gridDim.x) {
+    int c = (linearIndex / inputWidth / inputHeight) % numChannels;
+    int n = linearIndex / inputWidth / inputHeight / numChannels;
+    input += (n * numChannels + c) * outputHeight * outputWidth;
+    int maxind = indices[linearIndex];
+    output[linearIndex] = input[maxind];
+  }
+}
+
+template <typename T>
+__global__ void max_unpooling3d_backward_kernel(
+    T* gradOutputData,
+    int64_t oT,
+    int64_t oH,
+    int64_t oW,
+    PackedTensorAccessor<int64_t, 4> indices,
+    PackedTensorAccessor<T, 4> gradInput,
+    int offsetZ) {
+  int iColumn = blockIdx.x * blockDim.x + threadIdx.x;
+  int iRow = blockIdx.y * blockDim.y + threadIdx.y;
+  int iFrame = (blockIdx.z + offsetZ) % gradInput.size(1); // output frame/time
+  int slice =
+      (blockIdx.z + offsetZ) / gradInput.size(1); // output slice/feature
+
+  if (iRow < gradInput.size(2) && iColumn < gradInput.size(3)) {
+    int64_t index = indices[slice][iFrame][iRow][iColumn];
+    T grad_val = gradOutputData[slice * oT * oH * oW + index];
+    gradInput[slice][iFrame][iRow][iColumn] = grad_val;
+  }
+}
+
+Tensor& max_unpooling2d_forward_out_cuda(
+    Tensor& output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  AT_CHECK(
+      indices.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type Long");
+  auto owidth = output_size[0];
+  auto oheight = output_size[1];
+
+  TensorArg output_arg{output, "output", 1}, self_arg{self, "self", 2},
+      indices_arg{indices, "indices", 3};
+  checkAllSameGPU(
+      "max_unpooling2d_forward_out_cuda", {output_arg, self_arg, indices_arg});
+
+  AT_CHECK(self.numel() > 0, "Input must be non-empty tensor");
+
+  AT_CHECK(
+      (self.ndimension() == 3 || self.ndimension() == 4),
+      "Input to max_unpooling2d should be a 3d or 4d Tensor",
+      self.sizes());
+  AT_CHECK(
+      self.sizes() == indices.sizes(),
+      "Shape of input must match shape of indices");
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (width, height) in output_size");
+
+  int64_t dimw = 2;
+  int64_t dimh = 1;
+  int64_t numBatch = 1;
+
+  int64_t numChannels;
+  int64_t inputHeight;
+  int64_t inputWidth;
+
+  if (self.ndimension() == 4) {
+    numBatch = self.size(0);
+    dimw++;
+    dimh++;
+  }
+  numChannels = self.size(dimh - 1);
+  inputHeight = self.size(dimh);
+  inputWidth = self.size(dimw);
+
+  auto input_contiguous = self.contiguous();
+  auto indices_contiguous = indices.contiguous();
+
+  output.resize_({numBatch, numChannels, oheight, owidth});
+
+  output.zero_();
+
+  dim3 block(512);
+  dim3 grid((output.numel() + 512 - 1) / 512);
```

Comment:
I'm not sure where these constants popped out of, but we have inline functions like `GET_BLOCKS` for doing these and they should be preferred.

Here's what the old code used to look like:

```
  MaxUnpoolForward <<< GET_BLOCKS(count), CUDA_NUM_THREADS, 0, THCState_getCurrentStream(state) >>>
      (count, THCTensor_(data)(state, input), THCIndexTensor_(data)(state, indices),
      batchSize, nInputPlane, nInputRows, nInputCols, oheight, owidth, THCTensor_(data)(state, output));
  THCudaCheck(cudaGetLastError());
```